### PR TITLE
[feat]: [PL-21751]: Adds request filter to capture api auth scheme telemetry

### PR DESCRIPTION
**Pull request by ashishsanodia-harness**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)

### CODE REVIEW

File: 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  @Override
+  public void filter(ContainerRequestContext containerRequestContext) {
+    Optional<String> authorizationOptional = getAuthorizationFromHeaders(containerRequestContext);
+    Optional<String> apiKeyOptional = getApiKeyFromHeaders(containerRequestContext);
+    Optional<String> accountIdentifierOptional = getAccountIdentifierFromUri(containerRequestContext);
+    HashMap<String, Object> properties = new HashMap<>();
+
+    if (accountIdentifierOptional.isPresent()) {
```
Comment (vikas-harness) [inline 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java:61]: Why are we logging all events here?  What is the business need?

**Review by vikas-harness — CHANGES_REQUESTED** [review]

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)* status: Some of the checks are too old. 4 file(s) to review. 2 change requests. 0 of 2 approvals obtained. 

___
*[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 42 at r2](https://reviewable.io/reviews/harness/harness-core/30496#-MvHmv_w8YTfF0plArWY:-MvHmv_w8YTfF0plArWZ:boeahrf) ([raw file](https://github.com/harness/harness-core/blob/2bd8eade819088b667935c04d78c328e396275fc/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L42)):*
> ```Java
>   public static final String UNKNOWN = "Unknown";
>   public static final String AUTH_TYPE = "auth_type";
>   public static final String ACCOUNT_ID = "account_id";
> ```

should we make this `accountIdentifier` as this is NG keyword?
___
*[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 67 at r2](https://reviewable.io/reviews/harness/harness-core/30496#-MvHmmt13onmEpHEWgPq:-MvHmmt33RbM-bLxyJDT:b-6c7bqc) ([raw file](https://github.com/harness/harness-core/blob/2bd8eade819088b667935c04d78c328e396275fc/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L67)):*
> ```Java
>         properties.put(AUTH_TYPE, X_API_KEY);
>       } else {
>         properties.put(AUTH_TYPE, UNKNOWN);
> ```

should we add a log here to track such requests as this is not an expected case?

File: 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java (hunk 1 above)
Comment (ashishsanodia-harness) [reply 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java:61]: We want to see what all users are using automation(x-api-key) against our api. So here we are capturing three datapoints and sending it to segment
- account id
- api endpoint
- auth type

Comment (vikas-harness) [reply 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java:61]: @ashishsanodia-harness -- AFAIK, segment has a rate limit on the API so this level of data ingestion would overwhelm segment and downstream systems.

**Review by vikas-harness — CHANGES_REQUESTED** [review]

File: 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java (hunk 1 above)
Comment (akash-nagarajan) [reply 950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java:61]: @ashishsanodia-harness - In the below we should only send the event if the x-api-key header is set. Else we just ignore the whole thing

**Review by piyush-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)* status: 4 file(s) to review. 2 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by piyush-harness** [comment]
:lgtm:

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)* status: 4 file(s) to review. 2 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 95 at r3](https://reviewable.io/reviews/harness/harness-core/30496#-MvXJ7NA77FjSur_q1-B:-MvXJ7NA77FjSur_q1-C:br6fj8c) ([raw file](https://github.com/harness/harness-core/blob/f6eacc8db4b6b81c283b7add1f9a65a49a5a7c7c/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L95)):*
> ```Java
>       } catch (Exception exception) {
>         log.error("Error occurred while sending X-API-key telemetry data : account={}, endpoint={}, message={}",
>             properties.get(ACCOUNT_IDENTIFIER), properties.get(API_ENDPOINT), exception.getMessage());
> ```

Lets put complete exception in the error, and not only the message. In current scenario, we will lose out on exception stacktrace.

**Review by vikas-harness — DISMISSED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)* status: 4 file(s) to review. 1 change requests.

**Comment by ashishsanodia-harness** [comment]
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)_ status: Some of the checks are too old. 4 file(s) to review. 2 change requests. 0 of 2 approvals obtained.
> 
> _[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 42 at r2](https://reviewable.io/reviews/harness/harness-core/30496#-MvHmv_w8YTfF0plArWY:-MvHmv_w8YTfF0plArWZ:boeahrf) ([raw file](https://github.com/harness/harness-core/blob/2bd8eade819088b667935c04d78c328e396275fc/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L42)):_
> 
> > ```java
> >   public static final String UNKNOWN = "Unknown";
> >   public static final String AUTH_TYPE = "auth_type";
> >   public static final String ACCOUNT_ID = "account_id";
> > ```
> 
> should we make this `accountIdentifier` as this is NG keyword?
> >  Done
> 
> _[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 67 at r2](https://reviewable.io/reviews/harness/harness-core/30496#-MvHmmt13onmEpHEWgPq:-MvHmmt33RbM-bLxyJDT:b-6c7bqc) ([raw file](https://github.com/harness/harness-core/blob/2bd8eade819088b667935c04d78c328e396275fc/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L67)):_
> 
> > ```java
> >         properties.put(AUTH_TYPE, X_API_KEY);
> >       } else {
> >         properties.put(AUTH_TYPE, UNKNOWN);
> > ```
> 
> should we add a log here to track such requests as this is not an expected case?

> > Not needed anymore

**Comment by ashishsanodia-harness** [comment]
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)_ status: 4 file(s) to review. 2 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.
> 
> _[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 95 at r3](https://reviewable.io/reviews/harness/harness-core/30496#-MvXJ7NA77FjSur_q1-B:-MvXJ7NA77FjSur_q1-C:br6fj8c) ([raw file](https://github.com/harness/harness-core/blob/f6eacc8db4b6b81c283b7add1f9a65a49a5a7c7c/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L95)):_
> 
> > ```java
> >       } catch (Exception exception) {
> >         log.error("Error occurred while sending X-API-key telemetry data : account={}, endpoint={}, message={}",
> >             properties.get(ACCOUNT_IDENTIFIER), properties.get(API_ENDPOINT), exception.getMessage());
> > ```
> 
> Lets put complete exception in the error, and not only the message. In current scenario, we will lose out on exception stacktrace.

> > Done

**Review by ashishsanodia-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)* status: 4 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java, line 95 at r3](https://reviewable.io/reviews/harness/harness-core/30496#-MvXJ7NA77FjSur_q1-B:-MvXp71RC4odhwLzjOI5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f6eacc8db4b6b81c283b7add1f9a65a49a5a7c7c/950-telemetry/src/main/java/io/harness/telemetry/filter/APIAuthTelemetryFilter.java#L95)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Lets put complete exception in the error, and not only the message. In current scenario, we will lose out on exception stacktrace.
</blockquote></details>

Done.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r2, 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496#-:-MvXpuH28mt_qqINSkla:bnfp4nl)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by piyush-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30496)* status: :shipit: complete! ✓

**Pull request merged**